File: backend/schemas/donor.py
from pydantic import BaseModel, Field, field_validator
from typing import Optional
from enum import Enum
from fastapi import Query
from pydantic import BaseModel as PydanticBase

class LocationUpdate(PydanticBase):
    latitude:  float = Field(..., ge=-90,  le=90)
    longitude: float = Field(..., ge=-180, le=180)

class BloodGroup(str, Enum):
    A_POS  = "A+"
    A_NEG  = "A-"
    B_POS  = "B+"
    B_NEG  = "B-"
    AB_POS = "AB+"
    AB_NEG = "AB-"
    O_POS  = "O+"
    O_NEG  = "O-"


class AvailabilityStatus(str, Enum):
    AVAILABLE   = "available"
    UNAVAILABLE = "unavailable"
    BUSY        = "busy"


class DonorCreate(BaseModel):
    """Only donor-specific fields — no name/email/phone"""
    blood_group:  BloodGroup
    city:         str = Field(..., example="Mumbai")
    age:          int = Field(..., ge=18, le=65)
    availability: AvailabilityStatus = Field(default=AvailabilityStatus.AVAILABLE)


class DonorUpdate(BaseModel):
    city:         Optional[str] = None
    age:          Optional[int] = Field(None, ge=18, le=65)
    availability: Optional[AvailabilityStatus] = None


class DonorResponse(BaseModel):
    id:              int
    blood_group:     BloodGroup
    city:            str
    age:             int
    availability:    AvailabilityStatus
    is_active:       bool
    full_name:       str
    email:           str
    phone:           Optional[str]
    total_donations: int = 0
    lives_saved:     int = 0
    last_donation:   Optional[str] = None
    is_in_cooldown:  bool = False
    cooldown_until:  Optional[str] = None
    days_remaining:  int = 0
    # Location
    latitude:        Optional[float] = None
    longitude:       Optional[float] = None
    distance_km:     Optional[float] = None   # ← computed, not stored

    model_config = {"from_attributes": True}


class DonorListResponse(BaseModel):
    total:  int
    donors: list[DonorResponse]



class DonorFilterParams:
    def __init__(
        self,
        blood_group:  Optional[str]   = Query(default=None),
        city:         Optional[str]   = Query(default=None),
        is_available: Optional[bool]  = Query(default=None),
        search:       Optional[str]   = Query(default=None),
        # New location params
        lat:          Optional[float] = Query(default=None, description="Requester latitude"),
        lng:          Optional[float] = Query(default=None, description="Requester longitude"),
        radius_km:    Optional[float] = Query(default=None, description="Search radius in km"),
    ):
        self.blood_group  = blood_group
        self.city         = city
        self.is_available = is_available
        self.search       = search
        self.lat          = lat
        self.lng          = lng
        self.radius_km    = radius_km

File: App/lib/models/donor.dart
class Donor {
  final String id;
  final String fullName;
  final String email;
  final String bloodGroup;
  final String city;
  final int age;
  final bool isAvailable;
  final int totalDonations;
  final int livesSaved;
  final DateTime? lastDonation;

  // Cooldown fields — new
  final bool isInCooldown;
  final String? cooldownUntil;
  final int daysRemaining;

  Donor({
    required this.id,
    required this.fullName,
    required this.email,
    required this.bloodGroup,
    required this.city,
    required this.age,
    required this.isAvailable,
    this.totalDonations = 0,
    this.livesSaved = 0,
    this.lastDonation,
    this.isInCooldown = false,
    this.cooldownUntil,
    this.daysRemaining = 0,
  });

  factory Donor.fromJson(Map<String, dynamic> json) {
    bool parseAvailability(dynamic value) {
      if (value is bool) return value;
      if (value is String) return value.toLowerCase() == 'available';
      return true;
    }

    return Donor(
      id:             json['id']?.toString() ?? '',
      fullName:       json['full_name'] ?? '',
      email:          json['email'] ?? '',
      bloodGroup:     json['blood_group'] ?? '',
      city:           json['city'] ?? '',
      age:            json['age'] ?? 0,
      isAvailable:    parseAvailability(json['availability'] ?? json['is_available']),
      totalDonations: json['total_donations'] ?? 0,
      livesSaved:     json['lives_saved'] ?? 0,
      lastDonation:   json['last_donation'] != null
          ? DateTime.tryParse(json['last_donation'])
          : null,
      // Cooldown
      isInCooldown:   json['is_in_cooldown'] ?? false,
      cooldownUntil:  json['cooldown_until'],
      daysRemaining:  json['days_remaining'] ?? 0,
    );
  }

  // copyWith updated to include new fields
  Donor copyWith({
    String? id, String? fullName, String? email,
    String? bloodGroup, String? city, int? age,
    bool? isAvailable, int? totalDonations,
    int? livesSaved, DateTime? lastDonation,
    bool? isInCooldown, String? cooldownUntil, int? daysRemaining,
  }) {
    return Donor(
      id:             id ?? this.id,
      fullName:       fullName ?? this.fullName,
      email:          email ?? this.email,
      bloodGroup:     bloodGroup ?? this.bloodGroup,
      city:           city ?? this.city,
      age:            age ?? this.age,
      isAvailable:    isAvailable ?? this.isAvailable,
      totalDonations: totalDonations ?? this.totalDonations,
      livesSaved:     livesSaved ?? this.livesSaved,
      lastDonation:   lastDonation ?? this.lastDonation,
      isInCooldown:   isInCooldown ?? this.isInCooldown,
      cooldownUntil:  cooldownUntil ?? this.cooldownUntil,
      daysRemaining:  daysRemaining ?? this.daysRemaining,
    );
  }
}